kale/embed/mogonet.py

```diff
@@ -0,0 +1,121 @@
+import torch.nn.functional as F
+import torch_sparse
+from torch import Tensor
+from torch.nn import Module, Parameter
+from torch.nn.init import xavier_normal_
+from torch_geometric.nn.aggr import Aggregation
+from torch_geometric.nn.conv import MessagePassing
+from torch_sparse import SparseTensor
+
+
+class MogonetGCNConv(MessagePassing):
+    r"""Create message passing layers for MOGONET method.
+
+    Args:
+        in_channels (int): Size of each input sample.
+        out_channels (int): Size of each output sample.
+        bias (bool, optional): If set to :obj:`False`, the layer will not learn an additive bias. (default: :obj:`True`)
+        aggr (string or list or Aggregation, optional): The aggregation scheme to use,
+            *e.g.*, :obj:`"add"`, :obj:`"sum"` :obj:`"mean"`, :obj:`"min"`, :obj:`"max"` or :obj:`"mul"`.
+        **kwargs (optional): Additional arguments of :class:`torch_geometric.nn.conv.MessagePassing`.
+    """
+
+    def __init__(
+        self,
+        in_channels: int,
+        out_channels: int,
+        bias: bool = True,
+        aggr: Optional[Union[str, List[str], Aggregation]] = "add",
+        **kwargs,
+    ) -> None:
+        super().__init__(aggr=aggr, **kwargs)
+
+        self.in_channels = in_channels
+        self.out_channels = out_channels
+        self.weight = Parameter(torch.Tensor(self.in_channels, self.out_channels))
+
+        if bias:
+            self.bias = Parameter(torch.Tensor(self.out_channels))
+        else:
+            self.register_parameter("bias", None)
+
+        self.reset_parameters()
+
+    def reset_parameters(self) -> None:
+        """Initialize the parameters of the model."""
+        xavier_normal_(self.weight.data)
+
+        if self.bias is not None:
+            self.bias.data.fill_(0.0)
+
+    def forward(self, x: Tensor, edge_index: SparseTensor) -> Tensor:
+        # The format of edge_index is SparseTensor which performs fast sparse-matrix multiplication
+        x = torch.mm(x, self.weight)
+        out = self.propagate(edge_index, x=x)
+        return out
+
+    def message(self, x_j: Tensor) -> Tensor:
+        r"""Construct messages from node :math:`j` to node :math:`i` in analogy to :math:`\phi_{\mathbf{\Theta}}` for
+        each edge in :obj:`edge_index`.
+        """
+        return x_j
+
+    def message_and_aggregate(self, adj_t: Union[SparseTensor, Tensor], x: Tensor) -> Tensor:
+        r"""Fuse computations of :func:`message` and :func:`aggregate` into a single function."""
+        return torch_sparse.matmul(adj_t, x, reduce=self.aggr)
+
+    def update(self, aggr_out: Tensor) -> Tensor:
+        r"""Update node embeddings in analogy to :math:`\gamma_{\mathbf{\Theta}}` for each node
+        :math:`i \in \mathcal{V}`.
+        """
+        if self.bias is not None:
+            aggr_out = aggr_out + self.bias
+        return aggr_out
+
+
+class MogonetGCN(Module):
+    r"""Create the structure of the graph convolutional network in MOGONET method.
```

Comment:
There are already other GCN modules in PyKale. How specific is this to Mogonet? How different is this from others? Possible to unify? Otherwise, we may have many repeated modules.